apps/cowswap-frontend/src/modules/permit/utils/getPermitUtilsInstance.ts

```diff
@@ -0,0 +1,32 @@
+import { SupportedChainId } from '@cowprotocol/cow-sdk'
+import type { Web3Provider } from '@ethersproject/providers'
+
+import { Eip2612PermitUtils } from '@1inch/permit-signed-approvals-utils'
+
+import { PermitProviderConnector } from 'modules/wallet/utils/PermitProviderConnector'
+
+import { PERMIT_SIGNER } from '../const'
+
+const PERMIT_UTILS_CACHE: Record<number, Eip2612PermitUtils> = {}
+
+export function getPermitUtilsInstance(
+  chainId: SupportedChainId,
+  provider: Web3Provider,
+  account?: string | undefined
+): Eip2612PermitUtils {
+  const cached = PERMIT_UTILS_CACHE[chainId]
+
+  // Only cache if there's no account to prevent stale signers
+  if (!account && cached) {
```

Comment:
Afaiu the problem with caching utils for different provider instances?
Will it work if we do?:

```ts
const cache = new Map<Web3Provider, Eip2612PermitUtils>()

if (!cache[provider]) cache[provider] = eip2612PermitUtils

return cache[provider]
```

Reply:
I just don't fully understand, why don't we cache utils for accounts?

Reply:
I did not get to test this assumption, but the thinking is that the same account can be used with different providers.
Image connecting a Safe via WC or as a Safe app.

But maybe using the provider itself as the cache key as you propose would work?
I'll give that a try.

Reply:
Questions and potential issues of this:

1. Is the provider updated every time the account changes?
If it doesn't, we'll for sure get stale instances, which will fail to sign

2. For the "generic" case, if the provider updates when account changes, it'll create too many unnecessary instances
In the generic case, we don't care about the selected wallet, as we use the static PK to do the signing.
If caching by provider, we can have many unnecessary instances cached

Reply:
Updated 693cc25a9331baf67645183cd52a9e1119e4b2a8